### ISSUE TITLE
How to slice when multiple slice criterions exist in the program?

### ISSUE THREAD

**Issue by u88lx** [mchalupa/dg#354]
Hello,
Thanks for your shared tool dg. I am trying to slice a code taking a call of function as slicing criterion. When there are multiple function calls in the code, the result of the slicing is the union of each slice criterion. However, the desired result is supposed to be generated for only one function call point. For example:

```
void FUN1()
{
    int64_t * data;
    int64_t * dataBadBuffer = (int64_t *)ALLOCA(50*sizeof(int64_t));
    int64_t * dataGoodBuffer = (int64_t *)ALLOCA(100*sizeof(int64_t));
    data = dataBadBuffer;
    {
        int64_t source[100] = {0};
/* SLICE CRITERION*/
        memmove(data, source, 100*sizeof(int64_t));
    }
}
void FUNC2()
{
    int64_t * data;
    int64_t * dataBadBuffer = (int64_t *)ALLOCA(50*sizeof(int64_t));
    int64_t * dataGoodBuffer = (int64_t *)ALLOCA(100*sizeof(int64_t));
    data = dataGoodBuffer;
    {
        int64_t source[100] = {0}; 
/* SLICE CRITERION*/
        memmove(data, source, 100*sizeof(int64_t));
    }
}
int main(int argc, char * argv[])
{
    FUNC1();
    FUNC2();
    return 0;
}
```
If `memmove` is selected as a criterion, the slicing result will be the union of both 2 call of `memmove` in FUNC1 and FUNC2.  How could I obtain a slice only treats single call of `memmove` as a slice criterion?
Thanks in advance for helping with my confusion!

**Comment by u88lx** [comment]
Could an array be selected as a slice criterion? It was failed when an array or structure that was the argument of a function call was selected as a slice criterion. For example:
```
1        char source[100];
2        char dest[100] = "";
3        memset(source, 'A', 100-1);
4        source[100-1] = '\0';
5       if (data < 100)
6        {
7 /* If dest was selected as a slice criterion, dg will find no slice criterion*/
8           strncpy(dest, source, data);
9           dest[data] = '\0'; 
10         }
```
This code was compiled with command:
`clang -g -c -emit-llvm file.c`
`llvm-slicer -c 8:dest file.bc`
If `dest` was selected as a slice criterion, dg will find no slice criterion. How to use an array as a slice criterion?

**Comment by mchalupa** [comment]
Hi,

at this moment, there is no possibility of how to specify which call should be used as the slicing criterion. I plan to fix that soon, though. Meantime, you must e.g., rename the call or something like that.

> Could an array be selected as a slice criterion?

I think it should work, this is probably a bug. I'll check it out.

**Comment by mchalupa** [comment]
Yep, that is a bug. If you use '9:dest' instead, it works as expected, but it has problems with the strncpy...

**Comment by mchalupa** [comment]
Did the commit ff5d727 fix the issue it for you? (It worked for me...)

**Comment by mchalupa** [comment]
> How could I obtain a slice only treats single call of memmove as a slice criterion?

In the https://github.com/mchalupa/dg/tree/sc-impr branch, there is a patch that allows you to use also the line specification for calls, that is, you can use '-c 10:memmove' to specify which memmove to match. Try that please, if it works as you expect.

**Comment by u88lx** [comment]
Thanks for your help!
I hava tried both of the "[https://github.com/mchalupa/dg/tree/sc-impr]() branch" and the latest "master branch". And I was failed to solve both of my 2 prombles. In the latest "master branch",`No reachable slicing criteria: '8:dest'` was showed by dg, if `8:dest` was used to slicer. While `No reachable slicing criteria: '8:strncpy'` was showed by dg, if `8:strncpy` was used as a slice criterion. 
Both of the 2 branch were performed in docker. Is there any mistake I have made while using dg?
This is my testcase below:
[testcase_strncpy.zip](https://github.com/mchalupa/dg/files/5475550/testcase_strncpy.zip)

**Comment by mchalupa** [comment]
> This is my testcase below  

I see neither `dest` variable nor a call to strcpy on line 8 in any of those files...

**Comment by mchalupa** [comment]
ping

**Comment by mchalupa** [comment]
No activity in this issue for a while, closing. Feel free to re-open.
